Add semantics role for form

hannah-hyj opened the pull request:
fix: [#161628](https://github.com/flutter/flutter/issues/161628)
## Pre-launch Checklist



- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

chunhtai commented on packages/flutter/lib/src/widgets/form.dart:
```diff
@@ -280,18 +280,21 @@ class FormState extends State<Form> {
-      onWillPop: widget.onWillPop,
-      child: _FormScope(formState: this, generation: _generation, child: widget.child),
-    );
+    return Semantics(role: SemanticsRole.form, child: form);
```
This may need container: true so that it won't merge up and accidentally get other children to be included in the form

chunhtai replied:
e.g something like this may break without the container:true here
```dart
Semantics(container:true, child: Column(
  children: [Semantics(container: true, child: Text('1')), Form(...)]

))
```
you will se that text('1') somehow become a child in the form in the semantics tree

chunhtai commented on engine/src/flutter/lib/web_ui/lib/src/engine/semantics/form.dart:
```diff
@@ -0,0 +1,24 @@
+        semanticsObject,
+        preferredLabelRepresentation: LabelRepresentation.ariaLabel,
+      ) {
+    setAriaRole('form');
```
I vaguely remember web use the `<form>` tag to denote a form, and it is already used somewhere but for other purpose, cc @mdebbar , can you take a look to see if this is ok

hannah-hyj replied:
Thank you! updated to use the `<form>` element !

hannah-hyj commented on engine/src/flutter/lib/web_ui/test/engine/semantics/semantics_test.dart:
```diff
@@ -5399,6 +5402,30 @@ void _testLandmarks() {
   semantics().semanticsEnabled = false;
 }
 
+void _testForms() {
```
updated tests, thanks!

mdebbar reviewed (approved):
LGTM if the framework changes LGT @chunhtai

chunhtai reviewed (approved):
LGTM

hannah-hyj committed:
add form role

Update semantics_test.dart

Update licenses_flutter

Update packages/flutter/lib/src/widgets/form.dart

Update form.dart

Update engine/src/flutter/lib/web_ui/lib/src/engine/semantics/form.dart

Update engine/src/flutter/lib/web_ui/lib/src/engine/semantics/form.dart

Update form.dart

Update form.dart

add tests

lint

Add form role

Co-Authored-By: chunhtai <47866232+chunhtai@users.noreply.github.com>
Co-Authored-By: Mouad Debbar <mouad.debbar@gmail.com>
